macros/src/main/scala/pureconfig/Derivation.scala

```diff
@@ -48,17 +48,44 @@ class DerivationMacros(val c: whitebox.Context) extends LazyContextParser with M
     // but we may be unable to parse `Lazy` trees)
     val isHeadImplicitADerivation = c.openImplicits.headOption.exists(_.pre =:= typeOf[Derivation.type])
 
-    // check the `-Xmacro-settings:materialize-derivations` scalac flag
-    if (!isDerivationEnabled || !isHeadImplicitADerivation) {
-      // when not present, simply render the base `Derivation` constructor with `implcitly[A]`. This results in the same
-      // behavior as without `Derivation` (apart from the extra wrapper).
-      q"_root_.pureconfig.Derivation.Successful(_root_.scala.Predef.implicitly[${weakTypeOf[A]}])"
+    // check if the materialization was called explicitly, in which case we want to have the nicer compiler error
+    // messages
+    val isMaterializationExplicitCall = {
+      val firstMacroCall = c.enclosingMacros.reverse.find(ctx => ctx.prefix.tree.tpe =:= ctx.typeOf[Derivation.type])
+      firstMacroCall.fold(false)(_.openImplicits.isEmpty)
+    }
+
+    // Determine what to do if an implicit search fails. If we're in the context of an implicit search, we want to set
+    // the provided message in the @implicitNotFound annotation. If we're inside an explicit call to
+    // `materializeDerivation` we want to abort with the message
+    val onFailedImplicitSearch: String => Nothing = if (isMaterializationExplicitCall)
+      c.abort(c.enclosingPosition, _)
+    else
+      msg => {
+        setImplicitNotFound(msg)
+        // cause the implicit to fail materializing - the message is ignored
+        c.abort(c.enclosingPosition, "")
+      }
+
+    // check the `-Xmacro-settings:materialize-derivations` scalac flag and make sure we're not in an explicit
+    // materialization call
+    if ((!isDerivationEnabled || !isHeadImplicitADerivation) && !isMaterializationExplicitCall) {
+      // when not present, start an implicit search for `A` and place it inside a `Derivation.Successful` if the search
+      // succeeds.
+      val tpe = weakTypeOf[A]
+      inferImplicitValueCompat(tpe) match {
+        case EmptyTree => c.abort(c.enclosingPosition, "")
+        case t => q"_root_.pureconfig.Derivation.Successful[${tpe}]($t)"
+      }
 
     } else {
       // if `isRootDerivation` is `false`, then this is a `Derivation` triggered inside another `Derivation`
-      val isRootDerivation = c.openImplicits.count(_.pre =:= typeOf[Derivation.type]) == 1
+      val isRootDerivation = if (isMaterializationExplicitCall)
+        c.enclosingMacros.count(c => c.prefix.tree.tpe =:= c.typeOf[Derivation.type]) == 1
```

Comment:
Doesn't this check work for both explicit and implicit calls? I used `openImplicits` before because I was only considering implicit calls.

Reply:
I think we still need to have different checks. `enclosingMacros` is more like a stack trace for macros whereas `openImplicits` only shows implicit candidates currently being considered.

Consider the following example:

```scala
class C(x: Int, s: String)
case class B(c: C)
case class A(b: B)
implicitly[ConfigReader[A]]
```

By the time we're attempting to derive a `ConfigReader` for `B`, there will be two enclosing macros matching the check: the one for the derivation of `ConfigReader[A]` and another for the derivation of `ConfigReader[B]`. In this case, there will be only one open implicit: the one that attempts to derive a `MapShapedReader` for `B`. We want to consider both as root derivations to avoid deriving a `Failed` that is not collected by another derivation.

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
Hmm, I'm not sure I'm getting it... I don't think we ever want to consider two root derivations in the same implicit chain - only the outermost makes sense, since it's the only one who'll have access to the full AST.

In the case you mentioned it seems to be that there should be only one root derivation - the implicit search for `ConfigReader[A]`. If the search for a `ConfigReader[B]` fails it should still emit a `Failed` object, which would still be processed by the root derivation. Am I missing something? I have not looked into this code for a long time...

Reply:
I think that's only true for implicit searches that start on a derivation. In this case, `ConfigReader[A]` isn't wrapped in `Derivation`, so the entrypoint for the macro will be on `Derivation[ConfigReader[B]]`.

I think this exposes another underlying issue, though. I guess we were assuming that `isHeadImplicitADerivation` would be `false` in those cases, but it seems like the only way for it to be `false` is when `materializeDerivation` is called explicitly.

Reply:
Oh, I see what you mean; but `Derivation` is not supposed to work at all if the root implicit search is not of a `Derivation`.

Our macro relies on that because it sets the `@implicitNotFound` annotation on `Derivation` in order to show the custom message. Setting it on other classes would not be good since it is a "global" operation (i.e. other implicit searches in the same compilation unit would be affected by that) and I also believe that users shouldn't expect a custom message to appear if they are not explicitly requesting it.

Reply:
I think I finally found the bug! `openImplicits` is a stack, so the first implicit in the chain is not at the head, but rather at the tail of the list. Fixing the `isHeadImplicitADerivation` check allows us to only rely on the enclosing macros to check if we're in the root derivation, as you suggested.

Commit 99439e5 implements this fix.

Reply:
Great job debugging this!